codalab/lib/upload_manager.py

```diff
@@ -82,9 +121,100 @@ def _interpret_source(self, source: Source):
             filename = source[0]
         return is_url, is_fileobj, filename
 
-    def _unpack_fileobj(self, source_filename, source_fileobj, dest_path):
-        self.zip_util.unpack(
-            self.zip_util.get_archive_ext(source_filename), source_fileobj, dest_path
+
+class DiskStorageUploader(Uploader):
+    """Uploader that uploads to uncompressed files / folders on disk."""
+
+    @property
+    def storage_type(self):
+        return StorageType.DISK_STORAGE.value
+
+    def write_git_repo(self, source: str, bundle_path: str):
+        file_util.git_clone(source, bundle_path)
+
+    def write_fileobj(
+        self, source_ext: str, source_fileobj: IO[bytes], bundle_path: str, unpack_archive: bool
+    ):
+        if unpack_archive:
+            zip_util.unpack(source_ext, source_fileobj, bundle_path)
+        else:
+            with open(bundle_path, 'wb') as out:
+                shutil.copyfileobj(cast(IO, source_fileobj), out)
+
+
+class BlobStorageUploader(Uploader):
+    """Uploader that uploads to archive files + index files on Blob Storage."""
+
+    @property
+    def storage_type(self):
+        return StorageType.AZURE_BLOB_STORAGE.value
+
+    def write_git_repo(self, source: str, bundle_path: str):
+        with tempfile.TemporaryDirectory() as tmpdir:
+            file_util.git_clone(source, tmpdir)
+            # Upload a fileobj with the repo's tarred and gzipped contents.
+            self.write_fileobj(
+                ".tar.gz", tar_gzip_directory(tmpdir), bundle_path, unpack_archive=True
+            )
+
+    def write_fileobj(
+        self, source_ext: str, source_fileobj: IO[bytes], bundle_path: str, unpack_archive: bool
+    ):
+        if unpack_archive:
+            output_fileobj = zip_util.unpack_to_archive(source_ext, source_fileobj)
+        else:
+            output_fileobj = GzipStream(source_fileobj)
+        # Write archive file.
+        with FileSystems.create(bundle_path, compression_type=CompressionTypes.UNCOMPRESSED) as out:
```

Comment:
aren't archives compressed?

Reply:
Yes, but the fileobj we're writing to is already compressed. If we don't specify `compression_type=CompressionTypes.UNCOMPRESSED` to `FileSystems.create()`, Apache Beam will automatically _re-compress_ the given fileobj.

Reply:
makes sense